shared/src/balancer_event_handler.rs

```diff
@@ -167,3 +350,467 @@ fn convert_pool_registered(
+            (EventIndex::new(1, 0), token_registration_events[0].clone()),
+            // Next pool registered in block 1 with tokens only coming in block 2
+            // Not realistic, but we can handle it.
+            (EventIndex::new(1, 0), pool_registration_events[1].clone()),
+            (EventIndex::new(2, 0), token_registration_events[1].clone()),
+            // Next tokens registered in block 3, but corresponding pool not received till block 4
+            (EventIndex::new(3, 0), token_registration_events[2].clone()),
+            (EventIndex::new(4, 0), pool_registration_events[2].clone()),
+        ];
+
+        let mut pool_store = BalancerPools::default();
+        pool_store.insert_events(events).unwrap();
+        // Note that it is never expected that blocks for events will differ,
+        // but in this test block_created for the pool is the first block it receives.
+        assert_eq!(pool_store.last_event_block(), 3);
+        assert_eq!(
+            pool_store.pools_by_token.get(&tokens[0]).unwrap(),
+            &hashset! { pool_ids[0] }
+        );
+        assert_eq!(
+            pool_store.pools_by_token.get(&tokens[1]).unwrap(),
+            &hashset! { pool_ids[0], pool_ids[1] }
+        );
+        assert_eq!(
+            pool_store.pools_by_token.get(&tokens[2]).unwrap(),
+            &hashset! { pool_ids[1], pool_ids[2] }
+        );
+        assert_eq!(
+            pool_store.pools_by_token.get(&tokens[3]).unwrap(),
+            &hashset! { pool_ids[2] }
+        );
+
+        assert_eq!(
+            pool_store.pools.get(&pool_ids[0]).unwrap(),
+            &WeightedPool {
+                pool_id: pool_ids[0],
+                pool_address: pool_addresses[0],
+                tokens: vec![tokens[0], tokens[1]],
+                specialization: PoolSpecialization::new(0).unwrap(),
+                block_created: 1
+            }
+        );
+        assert_eq!(
+            pool_store.pools.get(&pool_ids[1]).unwrap(),
+            &WeightedPool {
+                pool_id: pool_ids[1],
+                pool_address: pool_addresses[1],
+                tokens: vec![tokens[1], tokens[2]],
+                specialization: PoolSpecialization::new(1).unwrap(),
+                block_created: 1
+            }
+        );
+        assert_eq!(
+            pool_store.pools.get(&pool_ids[2]).unwrap(),
+            &WeightedPool {
+                pool_id: pool_ids[2],
+                pool_address: pool_addresses[2],
+                tokens: vec![tokens[2], tokens[3]],
+                specialization: PoolSpecialization::new(2).unwrap(),
+                block_created: 3
+            }
+        );
+    }
+
+    struct BalancerPoolSetup {
+        pool_ids: Vec<H256>,
+        pool_addresses: Vec<H160>,
+        tokens: Vec<H160>,
+        specializations: Vec<PoolSpecialization>,
+        pool_store: BalancerPools,
+    }
+
+    fn dummy_balancer_setup(start_block: usize, end_block: usize) -> BalancerPoolSetup {
```

Comment:
This test scares me as a future maintainer of the codebase. It looks like there is a lot of undocumented setup which is reused in two test cases. If one of them fails I'm not sure I would understand what exactly happens in the setup and what to change to fix on but not the other test.

To me, a good unit test interface of this component would be the public interface, mainly `replace_events` and `append_events`. You could have some helper methods that create an event struct but the concrete "story" of which events are created in what order should be part of each individual test case. After processing the set of events you could make a bunch of assertions making sure the BalancerPools struct is in the expected state.

Reply:
I have rearranged, reduced, removed  and all-round cleaned up the test suite to be maintainable, accurate and exhaustive without duplication and only the public interfaces as suggested. Please let me know what you think!